좋아요 도메인 개발

haeuni00 opened the pull request:
https://devarium.atlassian.net/browse/BE-26

haeuni00 committed:
feat: Like 엔티티 구현

haeuni00 committed:
feat: Like 도메인 모델 추가

haeuni00 committed:
feat: LikeController 구현

haeuni00 committed:
feat: EntityType Enum 구현

- 좋아요를 할 수 있는 EntityType

haeuni00 committed:
feat: LikeService 인터페이스 구현

haeuni00 committed:
feat: LikeServiceImpl 구현

- 좋아요 추가 메서드 구현
- 좋아요 취소 메서드 구현

haeuni00 committed:
feat: LikeServiceDecorator 구현

haeuni00 committed:
feat: LikeRepository 인터페이스 구현

haeuni00 committed:
feat: LikeJpaRepository 인터페이스 구현

haeuni00 committed:
feat: LikeRepositoryImpl 구현

haeuni00 committed:
feat: LikeException 구현

- exception, errorCode 구현

haeuni00 committed:
feat: ServiceConfig에 Like 관련 bean 등록

haeuni00 committed:
feat: PostRepository에 exsitsById 추가

haeuni00 committed:
feat: CommentRepository에 exsitsById 추가

haeuni00 committed:
feat: ReplyRepository에 exsitsById 추가

haeuni00 committed:
chore: 메소드에 @Override 추가

haeuni00 committed:
feat: 좋아요 개수 조회 기능 구현

haeuni00 committed:
Merge branch 'dev' of https://github.com/devarium/devarium-backend into BE-26

# Conflicts:
#	infrastructure/src/main/java/io/devarium/infrastructure/persistence/repository/CommentRepositoryImpl.java
#	infrastructure/src/main/java/io/devarium/infrastructure/persistence/repository/PostRepositoryImpl.java
#	infrastructure/src/main/java/io/devarium/infrastructure/persistence/repository/ReplyRepositoryImpl.java

haeuni00 committed:
refactor: Post, Comment, Reply repositoryImpl 메서드 위치 변경

haeuni00 committed:
fix: LikeServiceImpl like 메소드 조건문 수정

haeuni00 committed:
feat: LikeException 핸들러 추가

sillysillyman reviewed (changes requested):
- `LikeController` 제거
  ```java
  @PostMapping("/{postId}/like")
  public ... likePost(
      @PathVariable Long postId,
      @AuthenticationPrincipal CustomUserPrincipal principal
  ) {
      ...
  }
  ```
  위와 같은 방식으로 각 도메인 컨트롤러에서 처리
- `Likeable` 인터페이스 정의
  - `Post`, `Comment`, `Reply`가 구현
  - `getId`, `getType` 메서드 정의
  - `LikeService`에서는 `Post` 등의 구현체가 아닌 `Likeable` 인터페이스로 비즈니스 로직 처리
- `LikeEntity` 컬럼 고유 제약 조건 추가
  ```java
  @Table(name = "likes",
      uniqueConstraints = @UniqueConstraint(
          columnNames = {"user_id", "target_type", "target_id"}
      )
  )
  ```

haeuni00 committed:
Merge branch 'dev' of https://github.com/devarium/devarium-backend into BE-26

# Conflicts:
#	api/src/main/java/io/devarium/api/common/GlobalExceptionHandler.java
#	infrastructure/src/main/java/io/devarium/infrastructure/persistence/config/ServiceConfig.java

haeuni00 committed:
feat: Likeable 인터페이스 구현

haeuni00 committed:
feat: LikeEntity uniqueConstraint 추가

haeuni00 committed:
feat: Post, Comment, Reply에 Likeable 인터페이스 구현

haeuni00 committed:
feat: PostController 좋아요 추가, 삭제, 조회 구현

haeuni00 committed:
feat: CommentController 좋아요 추가, 삭제, 조회 구현

haeuni00 committed:
feat: ReplyController 좋아요 추가, 삭제, 조회 구현

haeuni00 committed:
fix: merge 충돌 해결하다 빼먹은 import 추가

haeuni00 committed:
refactor: Like 도메인 비즈니스 로직 수정

sillysillyman commented on core/src/main/java/io/devarium/core/domain/comment/Comment.java:
```diff
@@ -36,4 +38,14 @@ public void validateAuthor(Long userId) {
         this.content = content;
     }
+
+    @Override
+    public Long getId() {
+        return this.id;
+    }
```
이건 `@Getter` 있어서 명시적으로 안 해줘도 되지 않나요? 확인 부탁드립니다.

haeuni00 replied:
앗 맞아요

sillysillyman commented on api/src/main/java/io/devarium/api/controller/reply/ReplyController.java:
```diff
@@ -68,4 +70,30 @@ public ResponseEntity<Void> deleteReply(
+        return ResponseEntity.noContent().build();
+    }
+
+    @GetMapping("/{replyId}/likes")
+    public ResponseEntity<Long> getReplyLikeCount(@PathVariable Long replyId) {
+        Reply reply = replyService.getReply(replyId);
+        return ResponseEntity.ok(likeService.getLikeCount(reply));
+    }
```
이것을 따로 API로 만든다기 보다는 `PostResponse`에 좋아요 수와 클릭 여부를 담는 것이 좋을 것 같습니다.
예를 들자면 아래와 같은 방식입니다.
```java
return PostResponse.of(
    post,
    likeService.getLikeCount(post),
    likeService.hasUserLiked(post, user)
);
```
좋아요 수와 클릭 여부는 따로 API를 호출하는 것보다 likeable 한 객체를 조회할 때 함께 전달하는 것이 좋을 것 같습니다.

haeuni00 replied:
Response에 from은 놔두고 좋아요 수와 클릭 여부를 담은 of 메서드를 따로 정의하는 방식인가요?

sillysillyman replied:
사실상 좋아요 수 없이 내용물만 조회할 일은 없을 것 같은데 Like 기능이 있는 이상 `from`이 불필요할 것 같긴 합니다.

sillysillyman commented on core/src/main/java/io/devarium/core/domain/like/Like.java:
core 모듈은 엔티티 개념과 독립적이기 때문에 `LikeableType`, `likableId` 또는 `targetId` 같은 것은 어떨까요?

haeuni00 replied:
좋은 것 같습니다

sillysillyman commented on core/src/main/java/io/devarium/core/domain/like/exception/LikeErrorCode.java:
```diff
@@ -0,0 +1,21 @@
+    LIKE_NOT_FOUND(HttpStatus.NOT_FOUND, "Like not found with id: %d"),
+    ENTITY_TYPE_NOT_FOUND(HttpStatus.NOT_FOUND, "EntityType not found with type: %s"),
+    ENTITY_ID_NOT_FOUND(HttpStatus.NOT_FOUND, "EntityId not found with id: %d")
+    ;
```
`;`를 윗줄로 옮겨주세요.

sillysillyman commented on core/src/main/java/io/devarium/core/domain/like/exception/LikeException.java:
```diff
@@ -0,0 +1,15 @@
+        super(errorCode.getMessage(args));
+        this.errorCode = errorCode;
+    }
+
```
공백 제거

sillysillyman commented on core/src/main/java/io/devarium/core/domain/like/service/LikeServiceImpl.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public void like(Likeable likeable, User user) {
+        if (likeRepository.existsByEntityTypeAndEntityIdAndUser(likeable.getType(), likeable.getId(), user)) {
```
`hasUserLiked`라는 메서드를 따로 정의하여 사용하는 게 좋을 것 같습니다. 이 조건문에서도 활용할 수 있고, 프론트엔드 입장에서도 로그인한 사용자가 좋아요를 눌렀는지 아닌지를 표시해줘야 하기 때문에 아마 `PostResponse`에 필요할 겁니다.

sillysillyman commented on infrastructure/src/main/java/io/devarium/infrastructure/persistence/config/ServiceConfig.java:
```diff
@@ -72,4 +76,13 @@ public TeamService teamService(TeamRepository teamRepository, MemberService memb
         return new MemberServiceDecorator(new MemberServiceImpl(memberRepository));
     }
+
+    @Bean
+    public LikeService likeService(
+        LikeRepository likeRepository, PostRepository postRepository,
+        CommentRepository commentRepository, ReplyRepository replyRepository) {
+        return new LikeServiceDecorator(
+            new LikeServiceImpl(likeRepository, postRepository, commentRepository,
+                replyRepository));
+    }
```
```java
@Bean
public LikeService likeService(
    LikeRepository likeRepository,
    PostRepository postRepository,
    CommentRepository commentRepository,
    ReplyRepository replyRepository
) {
    ....
}
```
매개변수가 많다면 이렇게 린팅해주세요.
그런데 `LikeServiceImpl`이 주입 받는 빈은 `LikeRepository` 하나인데 빈 4개를 주입하셔서 컴파일 에러 날 것 같아요.

haeuni00 replied:
여기를 정리 안했네요. 감사합니다

sillysillyman commented on infrastructure/src/main/java/io/devarium/infrastructure/persistence/repository/LikeRepositoryImpl.java:
```diff
@@ -0,0 +1,64 @@
+public class LikeRepositoryImpl implements LikeRepository {
+
+    private final LikeJpaRepository likeJpaRepository;
+    private final UserJpaRepository userJpaRepository;
```
리포지토리는 다른 리포지토리에 의존하면 안 됩니다. 다른 리포지토리 코드를 확인해주세요.

sillysillyman commented on infrastructure/src/main/java/io/devarium/infrastructure/persistence/repository/LikeRepositoryImpl.java:
```diff
@@ -0,0 +1,64 @@
+        likeJpaRepository.save(entity);
+    }
+
+    @Override
+    public boolean existsByEntityTypeAndEntityIdAndUser(EntityType entityType, Long typeId, User user) {
+        QLikeEntity like = QLikeEntity.likeEntity;
+        return queryFactory
+            .selectOne()
+            .from(like)
+            .where(like.entityType.eq(entityType), like.entityId.eq(typeId), like.user.id.eq(user.getId()))
+            .fetchFirst() != null;
+    }
```
추후에 쿼리 최적화 이슈에서 진행할 사항이긴 하지만, QueryDSL을 사용하신다면 repository 패키지 내에 querydsl 패키지를 생성하시고 해당 패키지에 `LikeQueryRepository`를 구현하시는 것은 어떨까요? 그리고 `LikeRepositoryImpl`은 `JPAQueryFactory`가 아닌 `LikeQueryRepository`를 의존하는 방식으로요.

haeuni00 replied:
이전에 다른 도메인 repositoryImpl에서 모두 QueryDSL을 사용해서 통일성을 맞추려던 것이었습니다. 현재는 모두 수정된 것으로 보여 이 부분은 통일성있게 수정하도록 하겠습니다.

haeuni00 committed:
refactor: getId 메서드 삭제

haeuni00 committed:
refactor: ServiceConfig likeService 매개변수 정리

haeuni00 committed:
refactor: EntityType, EntityId 필드명 변경

haeuni00 committed:
refactor: LikeService 유저가 이미 좋아요 했는지 확인 메서드 분리

haeuni00 committed:
refactor: exsitsById 메서드 삭제

haeuni00 committed:
refactor: Like 레포지토리 수정

haeuni00 committed:
refactor: 좋아요 개수, 클릭 여부 response에 포함

- Post, Comment, Reply 컨트롤러 좋아요 개수 메서드 삭제
- 각 response에 좋아요 개수, 클릭 여부 포함
- 각 컨트롤러에 reponse 사용 로직 수정

sillysillyman commented on api/src/main/java/io/devarium/api/controller/comment/dto/CommentResponse.java:
```diff
@@ -8,16 +8,20 @@ public record CommentResponse(
 ) {
 
-    public static CommentResponse from(Comment comment) {
+    public static CommentResponse from(Comment comment, Long likeCount, Boolean userLiked) {
```
매개변수가 여러개로 늘어났으니 `from` -> `of`로 메서드명 변경해주세요.

sillysillyman commented on api/src/main/java/io/devarium/api/controller/post/PostController.java:
```diff
@@ -34,24 +35,28 @@ public class PostController {
     @GetMapping("/{postId}")
-    public ResponseEntity<SingleItemResponse<PostResponse>> getPost(@PathVariable Long postId) {
+    public ResponseEntity<SingleItemResponse<PostResponse>> getPost(
+        @PathVariable Long postId,
+        @AuthenticationPrincipal CustomUserPrincipal principal
+    ) {
         Post post = postService.getPost(postId);
-        PostResponse response = PostResponse.from(post);
+        PostResponse response = createPostResponse(post, principal);
 
         return ResponseEntity.ok(SingleItemResponse.from(response));
     }
```
```java
@GetMapping("/{postId}")
public ResponseEntity<SingleItemResponse<PostResponse>> getPost(
    @PathVariable Long postId,
    @AuthenticationPrincipal(required = false) CustomUserPrincipal principal // 인증객체 null 허용, SecurityConfig에서 permitAll()
) {
    Post post = postService.getPost(postId);
    Long likeCount = likeService.getLikeCount(post);
    boolean liked = principal != null && likeService.hasUserLiked(post, principal.getUser()); // principal이 null이면 hasUserLiked 실행 안 됨
    // 로그인한 사용자 -> 좋아요 여부 포함
    // 비로그인 사용자 -> 좋아요 여부 무조건 false
    PostResponse response = PostResponse.of(post, likeCount, liked);

    return ResponseEntity.ok(SingleItemResponse.from(response));
}
```

이 코드를 활용해주세요. 반드시 이대로 해달라는 말씀은 아니고

1. 인증 객체가 없어도 되는 것 고려
2. 좋아요 여부는 true/false로 나눠주세요. 인증 객체가 없을 때도 null을 전달하는 것보다 false를 전달하는 게 좋을 것 같습니다.

당연히 이미 구현하신 대로 헬퍼메서드 정의해서 사용해주셔도 상관 없습니다. 비슷한 내용을 나눠서 적는 것보다 한번에 적느라고 풀어서 코드를 작성했습니다.

haeuni00 replied:
false로 전달하는거 좋은 것 같습니다.
SecurityConfig는 커밋을 안해서 안올라와있긴 하나 기존 코드에서 인증 객체가 없어도 돌아가는 것을 확인했고, 
@AuthenticationPrincipal에 required는 사용할 수 없는 것 같습니다.

sillysillyman reviewed (commented):
리포지토리 쪽만 리뷰되시면 merge 해도 될 것 같습니다.

haeuni00 committed:
refactor: 메서드명 변경

haeuni00 committed:
refactor: 페이징 메서드 객체 반환 -> reponse로 변환 로직 추가 및 변경(좋아요 개수와 클릭 여부 포함)

haeuni00 committed:
feat: SecurityConfig 요청 허용 추가
